File: odes/python/impulse.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Applied Numerical Analysis                                  August 25, 2021
ME 2020 FA21
Prof. M. Diaz-Maldonado


Synopsis:
Solves for the transient response of a first-order system when subjected
to an impulse-input.


Copyright (c) 2021 Misael Diaz-Maldonado
This file is released under the GNU General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.


References:
[0] A Gilat and V Subramanian, Numerical Methods for Engineers and
    Scientists: An Introduction with Applications using MATLAB
[1] R Johansson, Numerical Python: Scientific Computing and Data
    Science Applications with NumPy, SciPy, and Matplotlib, 2nd edition
[2] CA Kluever, Dynamic Systems: Modeling, Simulation, and Control
"""


from ode import iEuler
from ode import EulerRK2 as RK2
import numpy as np
from numpy import exp
from scipy.integrate import solve_ivp
import matplotlib as mpl
mpl.use("qt5agg")
from matplotlib import pyplot as plt


# initial value, and rate and forcing constants, respectively
yi, k, b = (0.0, 1.0, 1.0)
# defines the right-hand side RHS of the ODE: dy/dt = f(t, y) as a lambda
odefun = lambda t, y: (b - k * y)
# defines lambdas for the step and impulse-response for validation
step    = lambda t: (yi - b / k) * exp(-k * t) + b / k 
impulse = lambda t: -k * (yi - b / k) * exp(-k * t)


n = 255                         # number of integration time-steps
ti, tf = (0.0, 5.0)             # initial time ti and final time tf
tspan  = (ti, tf)               # time span tuple
odesol = np.empty([2, n+1, 2])  # preallocates array for speed


""" solves the ODE with Euler's and second-order Runge-Kutta Methods """
odesol[:, :, 0] = iEuler(n, tspan, yi, odefun)
odesol[:, :, 1] = RK2   (n, tspan, yi, odefun)
# uses scipy's implementation of the fourth-order Runge-Kutta method
odesol_scipy_IVP_Solver = solve_ivp(odefun, tspan, [yi], method="RK45")
# unpacks the numerical solutions
t, y_Euler, y_RK2 = (odesol[0, :, 0], odesol[1, :, 0], odesol[1, :, 1])
t_scipy, y_scipy  = (odesol_scipy_IVP_Solver.t, odesol_scipy_IVP_Solver.y)
y_scipy = y_scipy[0, :]

"""plots and compares the numerical solutions against the analytic ones"""

# step response
plt.close("all")            # closes all existing figures
plt.ion()                   # enables interactive plotting
fig, ax = plt.subplots()    # effectively plots on the same figure

ax.plot(t, step(t), color="black", linewidth=2.0,
        label="analytic solution")
ax.plot(t_scipy, y_scipy, color="blue", marker="v", linestyle="",
        label="scipy's IVP Solver")
ax.plot(t[::8], y_Euler[::8], color="orange", marker="o", linestyle="",
        label="implicit Euler's Method")
ax.plot(t[::16], y_RK2[::16], color="red",    marker="s", linestyle="",
        label="second-order Runge-Kutta Method")

ax.grid()                   # shows grid
ax.legend()                 # displays the legend
ax.set_xlabel("time, t")
ax.set_ylabel("dynamic response, y(t)")
ax.set_title("Step Response of a First Order Dynamic System")

# exports figure as a Portable Network Graphic PNG with 300 DPI resolution
fig.savefig("output/impulse/step.png", dpi=300)



# impulse response
fig, ax = plt.subplots()

ax.plot(t, impulse(t), color="black", linewidth=2.0,
        label="analytic solution")
ax.plot(t_scipy, odefun(t_scipy, y_scipy), color="blue", marker="v",
        linestyle="", label="scipy's IVP Solver")
ax.plot(t[::8], odefun(t[::8], y_Euler[::8]), color="orange", marker="o",
        linestyle="", label="implicit Euler's Method")
ax.plot(t[::16], odefun(t[::8], y_RK2[::16]), color="red",    marker="s",
        linestyle="", label="second-order Runge-Kutta Method")

ax.grid()
ax.legend()
ax.set_xlabel("time, t")
ax.set_ylabel("dynamic response, y(t)")
ax.set_title("Impulse Response of a First Order Dynamic System")
fig.savefig("output/impulse/impulse.png", dpi=300)

File: odes/cxx/impulse.cpp
/*
 * Applied Numerical Analysis			         September 20, 2021
 * ME 2020 FA21
 * Prof. M Diaz-Maldonado
 *
 * source: impulse.cpp
 *
 * Synopsis:
 * Obtains the transient response of a first-order Ordinary Differential
 * Equation ODE subject to step and impulse responses:
 *
 * 		y' + k * y = b * u(t),	y(0) = 0,
 * 
 * where k and b are the rate and forcing-constants, respectively, and
 * u(t) is either the unit-step or the unit-impulse function.
 *
 *
 * Copyright (c) 2021 Misael Diaz-Maldonado
 *
 * This file is released under the GNU General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 *
 * References:
 * [0] A Gilat and V Subramanian, Numerical Methods for Engineers and
 *     Scientists, 3rd edition.
 * [1] A Koenig and B Moo, Accelerated C++ Practical Programming by
 *     Example.
 * [2] www.cplusplus.com/reference/tuple/tuple
 * [3] www.geeksforgeeks.org/tuples-in-c/
 *
 */

#include <ios>
#include <cmath>
#include <tuple>
#include <vector>
#include <string>
#include <iomanip>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <functional>
// defines the rate and forcing constant, respectively
#define RATE 1.0
#define FEXT 1.0
// implements the absolute value function via MACROS
#define absval(x) (signbit(x)? -x: x)
import odes ;

using std::get ;
using std::cout ;
using std::endl ;
using std::tuple ;
using std::string ;
using std::vector ;
using std::signbit ;
using std::ofstream ;
using std::scientific ;
using std::streamsize ;
using std::setprecision ;
using std::runtime_error ;

using ode::iEuler ;
using ode::RK2 ;

// lambda, odefun, RHS of the ODE f(t, y)
std::function < double(double, double) >
odefun = [](double t, double y) -> double {
	const double k = RATE ;
	const double b = FEXT ;
	return (b - k * y) ;
} ;

// analytic expression for the step-response, y(t)
auto fstep = [](double t) -> double {
	const double k  = RATE ;
	const double b  = FEXT ;
	return ( -(b / k) * exp(-k * t) + b / k ) ;
} ;

// analytic expression for the impulse-response, y(t)
auto fimpulse = [](double t) -> double {
	const double k  = RATE ;
	const double b  = FEXT ;
	return  ( b * exp(-k * t) );
} ;

// loggers
void write (const std::tuple <std::vector<double>, std::vector<double>>& );
void write (const std::tuple < std::vector<double>, std::vector<double> >&,
	    const std::string& );

int main() {

	// defines parameters for the solver
	string filename ;
	const int N = 511 ;			// number of time-steps
	const double yi = 0.0 ;			// initial value y(0)
	const double ti = 0.0, tf = 3.0e1 ;	// time span

	// creates placeholders for the numerical solutions
	tuple < vector<double>, vector<double> > odesol_iEuler ;
	tuple < vector<double>, vector<double> > odesol_RK2 ;

	// solves the ODE using the specified method
	odesol_iEuler = iEuler (odesol_iEuler, ti, tf, yi, N, odefun);
	odesol_RK2    = RK2    (odesol_RK2,    ti, tf, yi, N, odefun);
	
	// exports the numerical solutions
	filename = "output/impulse/iEulr.dat" ;
	write (odesol_iEuler, filename);
	filename = "output/impulse/EuRK2.dat" ;
	write (odesol_RK2, filename);
	write (odesol_RK2);
	return 0 ;
}


void write (const tuple < vector<double>, vector<double> >& odesol)
{	// writes numerical results on the output stream

	// unpacks numerical solution from tuple
	const vector<double>& t = get<0>(odesol);
	const vector<double>& y = get<1>(odesol);

	streamsize prec = cout.precision() ;
	typedef std::vector<double>::size_type size ;
	for (size i = 0 ; i != t.size() ; ++i) {
		cout << scientific << setprecision(15) <<
			t[i] << "\t" << y[i] << "\t" <<
			odefun(t[i], y[i]) << "\t" <<
			absval( (fstep(t[i]) - y[i]) ) << "\t" <<
			absval( (fimpulse(t[i]) - odefun(t[i], y[i])) ) <<
			setprecision(prec) << endl ;
	}
}


void write (const tuple < vector<double>, vector<double> >& odesol,
	    const string& filename)
{	// writes numerical results to the output filestream

	const vector<double>& t = get<0>(odesol);
	const vector<double>& y = get<1>(odesol);

	ofstream out ;
	out.open (filename, std::ios::out);

	if ( !out.is_open() ) {
		throw runtime_error("I/O Error: " + filename);
	}

	streamsize prec = out.precision() ;
	typedef std::vector<double>::size_type size ;
	for (size i = 0 ; i != t.size() ; ++i) {
		out << scientific << setprecision(15) <<
			t[i] << "\t" << y[i] << "\t" <<
			odefun(t[i], y[i]) << "\t" <<
			absval( (fstep(t[i]) - y[i]) ) << "\t" <<
			absval( (fimpulse(t[i]) - odefun(t[i], y[i])) ) <<
			setprecision(prec) << endl ;
	}

	out.close() ;
}
